tools/cn-tree-simulator/cn/tree.go

```diff
@@ -0,0 +1,214 @@
+/* SPDX-License-Identifier: Apache-2.0
+ * Copyright (C) 2022 Micron Technology, Inc. All rights reserved.
+ */
+
+package cn
+
+import (
+	"container/list"
+	"encoding/json"
+	"fmt"
+)
+
+type Tree struct {
+	params *TreeParams
+	stats  treeStatistics
+
+	root rootNode
+}
+
+type node struct {
+	children *list.List
+	kvsets   *list.List
+}
+
+type rootNode struct {
+	node
+
+	params *RootNodeParams
+	stats  rootNodeStatistics
+}
+
+type leafNode struct {
+	node
+
+	params *LeafNodeParams
+	stats  leafNodeStatistics
+}
+
+func NewTree(leaves uint, initialSize uint64, params *Params) *Tree {
+	tree := Tree{
+		params: &params.Tree,
+		root: rootNode{
+			node: node{
+				children: list.New(),
+				kvsets:   list.New(),
+			},
+			params: &params.RootNode,
+		},
+	}
+
+	if initialSize != 0 {
+		for i := uint(0); i < leaves; i++ {
+			size := initialSize / uint64(leaves)
+			node := &leafNode{
+				node:   node{kvsets: list.New()},
+				params: &params.LeafNode,
+			}
+
+			tree.root.children.PushBack(node)
+
+			node.kvsets.PushFront(newKvSet(size, false))
+		}
+	}
+
+	return &tree
+}
+
+func (t *Tree) Size() uint64 {
+	total := t.root.size()
+
+	for leaf := t.root.children.Front(); leaf != nil; leaf = leaf.Next() {
+		for kvset := leaf.Value.(*leafNode).kvsets.Front(); kvset != nil; kvset = kvset.Next() {
+			total += kvset.Value.(*kvSet).size
+		}
+	}
+
+	return total
+}
+
+func (t *Tree) Ingest(data uint64) {
+	t.root.kvsets.PushFront(newKvSet(data, false))
+
+	t.stats.ingests++
+	t.stats.ingested += data
+	t.stats.written += data
+
+	if t.root.stats.spills == uint(0) && t.root.size() < t.root.params.InitialSpillSize {
+		return
+	}
+
+	if t.root.size() < t.root.params.SpillSize {
+		return
+	}
+
+	t.stats.written += t.root.size()
+	t.root.spill()
+
+	if t.stats.ingests%t.params.IngestionsTillKCompact == 0 {
+		t.kcompact()
+	} else {
+		return
+	}
+
+	if t.stats.ingests%t.params.IngestionsTillKVCompact == 0 {
+		t.kvCompact()
+	}
```

Comment:
I think a more natural code structure would be something like the following:
```
func (t *Tree) Ingest(data uint64) {
        t.root.ingest(data);
}
func (r *rootNode) Ingest(data uint64) {
	r.ingests++
	r.ingested += data
	r.written += data
	r.kvsets.PushFront(newKvSet(data, false))
        r.Maintain()
}
func (r *rootNode) Maintain() {
	if r.children.Len() == 0 {
                // apply rules for when root has no children
        } else {
                // apply rules for when root has children
                if .... {
                        r.spill()
                }
        }
}
func (r *rootNode) Spill() {
        r.spills += 1
	size := r.size() / uint64(r.children.Len())
	for leaf := r.children.Front(); leaf != nil; leaf = leaf.Next() {
                leaf.Ingest(size)
        }
}
func (leaf *leafNode) Ingest(data uint64) {
	leaf.ingests++
	leaf.ingested += data
	leaf.written += data
	leaf.kvsets.PushFront(newKvSet(data, false))
        leaf.Maintian()
}
func (leaf *leafNode) Maintain() {
        // apply k-compact, kv-compact, split rules  
}

```

Reply:
The example stores stats per node.   Could go through the nodes to get aggregate stats at end of test.

Might be easier in the long run to split "ingest" from "maintenance".  For example:  call ingest on each leaf node, which marks the nodes as "dirty" or puts them on a queue of nodes that might need work.  And then call "maintenance" on each dirty node on the way out (or at the end of the event loop).

Reply:
Sure. Let me take a stab at a little re-org. The maintain method seems like a good change, and ingesting through the root node sounds like a good change theoretically speaking in order to be similar to HSE.